crates/evm/src/evm/system_events.rs

```diff
@@ -0,0 +1,80 @@
+use reth_primitives::{
+    address, Address, Bytes as RethBytes, Signature, Transaction, TransactionKind,
+    TransactionSigned, TransactionSignedEcRecovered, TransactionSignedNoHash, TxEip1559, U256,
+};
+
+use super::system_contracts::L1BlockHashList;
+
+/// This is a special signature to force tx.signer to be set to SYSTEM_SIGNER
+pub const SYSTEM_SIGNATURE: Signature = Signature {
+    r: U256::ZERO,
+    s: U256::ZERO,
+    odd_y_parity: false,
+};
+
+/// This is a special system address to indicate a tx is called by system not by a user/contract.
+pub const SYSTEM_SIGNER: Address = address!("deaddeaddeaddeaddeaddeaddeaddeaddeaddead");
+
+/// A system event is an event that is emitted on special conditions by the EVM.
+/// There events will be transformed into Evm transactions and put in the begining of the block.
+#[derive(Clone, Debug, serde::Serialize, serde::Deserialize, Eq, PartialEq)]
+pub(crate) enum SystemEvent {
+    L1BlockHashInitialize(/*block number*/ u64),
+    L1BlockHashSetBlockInfo(/*hash*/ [u8; 32], /*merkle root*/ [u8; 32]),
+}
+
+fn system_event_to_transaction(event: SystemEvent, nonce: u64, chain_id: u64) -> Transaction {
+    let sys_block_hash = L1BlockHashList::default();
+    let body: TxEip1559 = match event {
+        SystemEvent::L1BlockHashInitialize(block_number) => TxEip1559 {
+            to: TransactionKind::Call(L1BlockHashList::address()),
+            input: RethBytes::from(sys_block_hash.init(block_number).to_vec()),
+            nonce,
+            chain_id,
+            value: U256::ZERO,
+            gas_limit: 1_000_000u64,
+            max_fee_per_gas: u64::MAX as u128,
+            ..Default::default()
+        },
+        SystemEvent::L1BlockHashSetBlockInfo(block_hash, txs_commitments) => TxEip1559 {
+            to: TransactionKind::Call(L1BlockHashList::address()),
+            input: RethBytes::from(
+                sys_block_hash
+                    .set_block_info(block_hash, txs_commitments)
+                    .to_vec(),
+            ),
+            nonce,
+            chain_id,
+            value: U256::ZERO,
+            gas_limit: 1_000_000u64,
+            max_fee_per_gas: u64::MAX as u128,
```

Comment:
what about max fee per gas 0, just to be clear to the reader.

Reply:
Then I will get Transaction(GasPriceLessThanBasefee) error, because we check gas fees for system txs.